src/collapse/perf.rs

```diff
@@ -150,33 +205,202 @@ impl Collapse for Folder {
+            // therefore we set up a channel to be able to communicate what the event filter state
+            // should be as soon as we know it and make the first thing each thread does be waiting
+            // on receiving information from this channel.
+            let (tx_event_filter, rx_event_filter) = channel::bounded(self.opt.nthreads);
+
+            // If we know what the event filter state should be without having to look at the data
+            // (because the user provided it), send this information down the channel immediately.
+            if self.event_filter.is_some() {
+                for _ in 0..self.opt.nthreads {
+                    tx_event_filter.send(self.event_filter.clone()).unwrap();
+                }
+            }
+
+            // Spin up the threadpool / worker threads.
+            let (tx_input, rx_input) = channel::bounded::<Option<Vec<u8>>>(2 * self.opt.nthreads);
+            let (tx_output, rx_output) = channel::bounded(self.opt.nthreads);
+            let mut handles = Vec::with_capacity(self.opt.nthreads);
+            for _ in 0..self.opt.nthreads {
+                let rx_event_filter = rx_event_filter.clone();
+                let rx_input = rx_input.clone();
+                let tx_output = tx_output.clone();
+
+                let nstacks_per_job = self.nstacks_per_job;
+                let occurrences = self.occurrences.clone();
+                let opt = self.opt.clone();
+
+                let handle = scope.spawn(move |_| {
+                    // Don't do anything until we know what the event filter should be
+                    let event_filter = rx_event_filter.recv().unwrap();
+                    let mut folder = Folder {
+                        cache_line: Vec::default(),
+                        event_filter,
+                        in_event: false,
+                        nstacks_per_job,
+                        occurrences,
+                        pname: String::default(),
+                        skip_stack: false,
+                        stack: VecDeque::default(),
+                        opt,
+                    };
+                    let mut sent_output = false;
+                    loop {
+                        if let Some(data) = rx_input.recv().unwrap() {
+                            if !sent_output {
+                                if let Err(e) = folder.collapse_single_threaded(&data[..]) {
+                                    tx_output.send(Err(e)).unwrap();
+                                    sent_output = true;
+                                    continue;
+                                }
+                                folder.in_event = false;
+                                folder.pname.clear();
+                                folder.skip_stack = false;
+                                folder.stack.clear();
+                            }
+                        } else {
+                            tx_output.send(Ok(())).unwrap();
```

Comment:
You probably don't even need this. When the threads exit, they'll all drop their `tx` handles, so the receiver's `.recv()` will return any errors followed by `None` as the channel has been closed.